### PR TITLE
adding support to return logits and generate for Megatron-LM GPT models

### LINKED ISSUE #809: [Bug] importlib.metadata.PackageNotFoundError: megatron-lm
### System Info

```Shell
p.s. `accelerate env` crashes.

You only need the latest accelerate commit b816e258a95904736c8e3a5a2b14a18d5f407f4a.
```

### Information

- [X] The official example scripts
- [ ] My own modified scripts

### Tasks

- [X] One of the scripts in the examples/ folder of Accelerate or an officially supported `no_trainer` script in the `examples` folder of the `transformers` repo (such as `run_no_trainer_glue.py`)
- [ ] My own task or dataset (give details below)

### Reproduction

1. Clone `Megatron-LM`.
2. Add the path of `Megatron-LM` to `PYTHONPATH`.
3. `import accelerate`.

### Expected behavior

```Shell
`accelerate` should not crash.

With https://github.com/huggingface/accelerate/pull/755, accelerate checks whether `megatron` is available.
However, once it is available, accelerate loads `megatron-lm`.

Error message:

Traceback (most recent call last):
  File "/home/ubuntu/workspace_hf/transformers/src/transformers/utils/import_utils.py", line 1063, in _get_module
    return importlib.import_module("." + module_name, self.__name__)
  File "/home/ubuntu/anaconda3/envs/hf-bench-env/lib/python3.8/importlib/__init__.py", line 127, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "<frozen importlib._bootstrap>", line 1014, in _gcd_import
  File "<frozen importlib._bootstrap>", line 991, in _find_and_load
  File "<frozen importlib._bootstrap>", line 975, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 671, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 843, in exec_module
  File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
  File "/home/ubuntu/workspace_hf/transformers/src/transformers/modeling_utils.py", line 39, in <module>
    from .deepspeed import deepspeed_config, is_deepspeed_zero3_enabled
  File "/home/ubuntu/workspace_hf/transformers/src/transformers/deepspeed.py", line 38, in <module>
    from accelerate.utils.deepspeed import HfDeepSpeedConfig as DeepSpeedConfig
  File "/home/ubuntu/workspace_hf/accelerate/src/accelerate/__init__.py", line 7, in <module>
    from .accelerator import Accelerator
  File "/home/ubuntu/workspace_hf/accelerate/src/accelerate/accelerator.py", line 27, in <module>
    from .checkpointing import load_accelerator_state, load_custom_state, save_accelerator_state, save_custom_state
  File "/home/ubuntu/workspace_hf/accelerate/src/accelerate/checkpointing.py", line 24, in <module>
    from .utils import (
  File "/home/ubuntu/workspace_hf/accelerate/src/accelerate/utils/__init__.py", line 100, in <module>
    from .megatron_lm import (
  File "/home/ubuntu/workspace_hf/accelerate/src/accelerate/utils/megatron_lm.py", line 38, in <module>
    if is_megatron_lm_available():
  File "/home/ubuntu/workspace_hf/accelerate/src/accelerate/utils/imports.py", line 96, in is_megatron_lm_available
    megatron_version = parse(importlib_metadata.version("megatron-lm"))
  File "/home/ubuntu/anaconda3/envs/hf-bench-env/lib/python3.8/importlib/metadata.py", line 530, in version
    return distribution(distribution_name).version
  File "/home/ubuntu/anaconda3/envs/hf-bench-env/lib/python3.8/importlib/metadata.py", line 503, in distribution
    return Distribution.from_name(distribution_name)
  File "/home/ubuntu/anaconda3/envs/hf-bench-env/lib/python3.8/importlib/metadata.py", line 177, in from_name
    raise PackageNotFoundError(name)
importlib.metadata.PackageNotFoundError: megatron-lm
```
```

### PR DESCRIPTION
**Pull request by pacman100**
### What does this PR do?
1. Enable returning logits for Megatron-LM GPT models
2. Enable specify custom Megatron-LM based model 
3. Enable specify any Megatron-LM arguments to support things such as ALiBi/ROPE positional embeddings, Multi-Query Attention, tokenizer files etc.
4. Enable `megatron_generate` method for Megatron-LM GPT model, this will use Tensor and Pipeline Parallelism to complete generations for a batch of `inputs` when using greedy with/without top_k/top_p sampling  and for individual prompt `inputs` when using beam search decoding. Only a subset of features of transformers `generate` is supported. This will help in using large models via tensor and pipeline parallelism  for generation (already does key-value caching and uses fused kernels by default). 
5. fixes #809

Below is the run of the example script [megatron_gpt2_generation.py](https://github.com/pacman100/accelerate-megatron-test/blob/main/src/inference/megatron_gpt2_generation.py) with the main parts of output logs given below:
```
The Megatron LM model weights are initialized at random in `accelerator.prepare`. Please use `accelerator.load_checkpoint` to load a pre-trained checkpoint matching the distributed setup.
 > number of parameters on (tensor, pipeline) model parallel rank (0, 0): 102483968
 > number of parameters on (tensor, pipeline) model parallel rank (0, 1): 101437440
 > number of parameters on (tensor, pipeline) model parallel rank (1, 0): 102483968
 > number of parameters on (tensor, pipeline) model parallel rank (1, 1): 101437440
Preparing optimizer
> learning rate decay style: linear
Preparing scheduler
...

11/04/2022 14:47:18 - INFO - __main__ - ***** Running training *****
11/04/2022 14:47:18 - INFO - __main__ -   Num examples = 2318
11/04/2022 14:47:18 - INFO - __main__ -   Num Epochs = 2
11/04/2022 14:47:18 - INFO - __main__ -   Instantaneous batch size per device = 24
11/04/2022 14:47:18 - INFO - __main__ -   Total train batch size (w. parallel, distributed & accumulation) = 48
11/04/2022 14:47:18 - INFO - __main__ -   Gradient Accumulation steps = 1
11/04/2022 14:47:18 - INFO - __main__ -   Total optimization steps = 96
  0%|                                                                        | 0/96 [00:00<?, ?it/s]
Resumed from checkpoint: /home/sourab/temp/megatron_lm_checkpoint
11/04/2022 14:47:18 - INFO - accelerate.accelerator - Loading states from /home/sourab/temp/megatron_lm_checkpoint
11/04/2022 14:47:18 - INFO - accelerate.accelerator - Loading Megatron-LM Model, Optimizer and Scheduler
Resuming from /home/sourab/temp/megatron_lm_checkpoint
 loading release checkpoint from /home/sourab/temp/megatron_lm_checkpoint
  Warning, trying to load an old checkpoint: 'types.SimpleNamespace' object has no attribute 'position_embedding_type'
 checkpoint version 3.0
  successfully loaded checkpoint from /home/sourab/temp/megatron_lm_checkpoint at iteration 0
11/04/2022 14:47:18 - INFO - accelerate.accelerator - Megatron-LM Model , Optimizer and Scheduler loaded from input dir /home/sourab/temp/megatron_lm_checkpoint
11/04/2022 14:47:18 - INFO - accelerate.checkpointing - All model weights loaded successfully
11/04/2022 14:47:18 - INFO - accelerate.checkpointing - All optimizer states loaded successfully
11/04/2022 14:47:18 - INFO - accelerate.checkpointing - All scheduler states loaded successfully
11/04/2022 14:47:18 - INFO - accelerate.checkpointing - Could not load random states
11/04/2022 14:47:18 - INFO - accelerate.accelerator - Loading in 0 custom states
  0%|                                                                        | 0/96 [00:00<?, ?it/s]
accelerator.process_index=3 outputs.logits=tensor([[[ 4.4648,  3.7715, -3.1172,  ...,  1.4053,  1.4053,  1.4053],
         [ 0.6279,  3.5449, -1.7275,  ...,  0.9048,  0.9048,  0.9048],
         [ 3.1074,  5.1484, -2.4570,  ...,  3.0684,  3.0684,  3.0684],
         ...,
       device='cuda:3', grad_fn=<CatBackward0>)
 50%|███████████████████████████████▌                               | 48/96 [00:40<00:36,  1.33it/s]11/04/2022 14:48:00 - INFO - __main__ - epoch 0: perplexity: 14.762350004624928 eval_loss: 2.692080020904541
epoch 0 training + evaluation took 0.708652 minutes
100%|███████████████████████████████████████████████████████████████| 96/96 [01:19<00:00,  1.32it/s]11/04/2022 14:48:39 - INFO - __main__ - epoch 1: perplexity: 14.325097285606006 eval_loss: 2.662013053894043
epoch 1 training + evaluation took 0.648274 minutes
Total Training + Evaluation took 1.356926 minutes

...
Greedy generation on a batch of 4 below with adding `bos` token at the start
['<|endoftext|>Are you human or are you... conventional?: What the heck is being called ` `... from television\'together? * Smiles * " "\' ( BBC ) ” [ 64 Bill Atkinson, Ep. 77 ], the stage play\'[ 61 G.A.U.S.L.D.\'], based on Jack Lee\'s Animal Volumes \'',

 '<|endoftext|>The purpose of life is to make you think, not to glorify. I try to live as I want to live ; I live at broke bread and frugal with my excess. There are times when you might not like what you see painted on a work of art, but you have to accept it when you have the opportunity to see the full light of happiness.', 

'<|endoftext|>The arsenal was constructed at the request of Germany in the mid @-@ 1930s, first in the Frankfurt @-@ Burgess Conservatory for the Erkenau Expedition and then in the Kattegat Protectorate, unused since Eastern Front. About twenty guns were taken from the Levant and transferred to the modern arsenal. The first Heinkel Bf 110 dive bombers were', 

'<|endoftext|>How are you doing these days? Last night you were really really quiet for a while, no mean feat for a baby of your age. Has that dog been around for years? " How are you doing, Baby? That is a really interesting question, " says my daughter, matter of factly, actually giving me the space to actually reassure her that everything and everyone is']

Beam generation on an individual input below
['The purpose of life is to make the world a better place. " \n = = = Education = = = \n = = = = Education = = = = \n = = = = Education = = = = \n = = = = Education = = = = \n = = = = Education = = = = \n = = =']
100%|███████████████████████████████████████████████████████████████| 96/96 [01:30<00:00,  1.06it/s]
wandb: Waiting for W&B process to finish... (success).
wandb:                                                                                
wandb: 
wandb: Run history:
wandb:      epoch ▁█
wandb:  eval_loss █▁
wandb: perplexity █▁
wandb:       step ▁█
wandb: train_loss █▁
wandb: 
wandb: Run summary:
wandb:      epoch 1
wandb:  eval_loss 2.66201
wandb: perplexity 14.3251
wandb:       step 96
wandb: train_loss 2.68363
```

### CODE REVIEW

File: src/accelerate/utils/imports.py [hunk 1]
```diff
@@ -90,10 +91,11 @@ def is_bf16_available(ignore_tpu=False):
 
 
 def is_megatron_lm_available():
-    package_exists = importlib.util.find_spec("megatron") is not None
-    if package_exists:
-        megatron_version = parse(importlib_metadata.version("megatron-lm"))
-        return compare_versions(megatron_version, ">=", "2.2.0")
+    if os.environ.get("USE_MEGATRON_LM", "false") == "true":
```
Comment (sgugger) [inline src/accelerate/utils/imports.py:94]: Here you should lowercase, or use the str2bool conversion, just to make sure to catch variants.

File: src/accelerate/utils/megatron_lm.py [hunk 2]
```diff
@@ -1134,10 +1171,18 @@ def forward(self, **batch_data):
...
+        logits = None
+        if "logits" in loss_dict:
+            logits = loss_dict["logits"]
         # loss = reduce(loss)
         if self.train_step_handler.model_output_class is not None:
-            return self.train_step_handler.model_output_class(loss=loss)
+            return self.train_step_handler.model_output_class(
+                loss=loss,
+                logits=logits,
+            )
```
Comment (sgugger) [inline src/accelerate/utils/megatron_lm.py:1185]: Fits on one line.

File: src/accelerate/utils/megatron_lm.py [hunk 3]
```diff
@@ -1186,6 +1231,184 @@ def load_checkpoint(self, input_dir):
         if args.fp16 and self.iteration == 0:
             self.optimizer.reload_model_params()
 
+    def generate(
```
Comment (sgugger) [inline src/accelerate/utils/megatron_lm.py:1234]: Very reluctant to add this here with the same name as in Transformers. This method does not have the same signature/behavior so users will be very confused to get something different.

**Review by sgugger — COMMENTED** [review]
Thanks for working on this! I think the `generate` method you are adding should be renamed to something including the name `meagtron`, otherwise we will confuse users that might expect to get the Transformers generate method and all it supports.

File: src/accelerate/utils/imports.py (hunk 1 above)
Comment (pacman100) [reply src/accelerate/utils/imports.py]: Done.

File: src/accelerate/utils/megatron_lm.py (hunk 2 above)
Comment (pacman100) [reply src/accelerate/utils/megatron_lm.py]: Done.

File: src/accelerate/utils/megatron_lm.py (hunk 3 above)
Comment (pacman100) [reply src/accelerate/utils/megatron_lm.py]: renamed to `megatron_generate`

**Review by sgugger — APPROVED** [review]

**Review by sgugger — APPROVED** [review]
Thanks for adding this!

**Pull request merged**